### PR TITLE
update rustc 1.9.0-nightly (d5a91e695 2016-03-26)

### LINKED ISSUE #805: Clippy broken on rustc 1.9.0-nightly (d5a91e695 2016-03-26)
See the change here https://github.com/rust-lang/rust/commit/5efdde0de161a5b374e16e20d09322e5a0eba4f9

I have a fix nearly ready will submit within the next hour

### PR DESCRIPTION
**Pull request by josephDunne**
move cfg, infer, traits and ty from middle to top-level
move middle::subst into middle::ty
track the extern-crate def-id rather than path (rustc ab9b844)

fixes #805

### CODE REVIEW

**Comment by josephDunne** [comment]
rust-lang/rust#32535 is preventing the tests from passing. This is just a preview pull request for your info

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -31,7 +31,7 @@ the lint will end up to be a nested series of matches and ifs,
 [like so](https://github.com/Manishearth/rust-clippy/blob/de5ccdfab68a5e37689f3c950ed1532ba9d652a0/src/misc.rs#L34).
 
 T-middle issues can be more involved and require verifying types. The
-[`middle::ty`](http://manishearth.github.io/rust-internals-docs/rustc/middle/ty) module contains a
+[`ty`](http://manishearth.github.io/rust-internals-docs/rustc/middle/ty) module contains a
```
Comment (Manishearth) [inline CONTRIBUTING.md]: link needs updating too

File: src/enum_glob_use.rs [hunk 2]
```diff
@@ -51,8 +51,9 @@ impl EnumGlobUse {
                             }
                         }
                     } else {
-                        if let Some(dp) = cx.sess().cstore.def_path(def.def_id()).last() {
-                            if let DefPathData::Type(_) = dp.data {
+                        let dp = cx.sess().cstore.relative_def_path(def.def_id());
+                        if  let Some(dpa) = dp.data.last() {
+                            if let  DefPathData::TypeNs(_) = dpa.data {
```
Comment (eddyb) [inline src/enum_glob_use.rs]: The change from `Type` to `TypeNs` is likely wrong - IIRC `TypeNs` is also used for modules, but I need to check.

Comment (mcarton) [inline src/enum_glob_use.rs]: Looks like `Type` and `Mod` were merged into `TypeNs` here, you must differentiate the two cases somehow.

Comment (eddyb) [inline src/enum_glob_use.rs]: Also, I think this whole thing is bogus, why are you not just checking for `Def::Enum` in the first place?

Comment (Manishearth) [inline src/enum_glob_use.rs]: glob imports are Def::Mod for some reason

**Comment by Manishearth** [comment]
Fixed it up and published. Thanks for the help!

**Comment by josephDunne** [comment]
Awesome :-) for my education can you give a quick overview of what that code was trying to do and why you replaced it with what you did

**Comment by Manishearth** [comment]
It was using a part of the metadata that was meant for incremental compilation, which isn't reliable for what we wanted. Now, it instead ensures that the definition contains a variant as its first child.

**Pull request merged**